Support for default item in match.filter_by

fungus opened the pull request:
### What does this PR do?
Adds a default parameter to the match.filter_by function.  This allows custom return for no matches found.
### What issues does this PR fix or reference?
#23050
### Previous Behavior
Returns None on no matches.

### Tests written?
No

fungus committed:
Support for default item in match.filter_by #23050

salt-jenkins commented:
@fungus, thanks for your PR! By analyzing the history of the files in this pull request, we identified @whiteinge, @terminalmage and @thatch45 to be potential reviewers.

whiteinge commented on salt/modules/match.py:
```diff
@@ -329,7 +330,7 @@ def filter_by(lookup,
             'web*': ['app', 'caching'],
             'db*': ['db'],
-        }) %}
+        }, default=['other']) %}
```
`default='web*'` perhaps?

whiteinge commented on salt/modules/match.py:
```diff
@@ -353,7 +354,7 @@ def filter_by(lookup,
             return lookup[key]
 
-    return None
+    return default
```
Should be `lookup.get(default, None)`

whiteinge commented on salt/modules/match.py:
```diff
@@ -309,7 +309,8 @@ def glob(tgt, minion_id=None):
               minion_id=None,
-              expr_form=None):
+              expr_form=None,
+              default=None):
```
I'd suggest using `default='default'` to match the grains function.

whiteinge reviewed (changes requested):
Sorely needed addition. Thanks for diving in and getting it added! Just a few notes.

fungus committed:
Change behavior of default parameter

Match behavior of default parameter in grains.filter_by.  Specify the default key in the lookup table, instead of specifying the default value.

fungus commented:
Got it.  I thought it made more sense to specify a default value, instead of a default key in the lookup table.  However, matching the behavior of grains.filter_by is more important.  Commit added.

whiteinge commented:
Thank you!

cachedout commented:
Thank you very much, @fungus
